interpreter/cling/lib/Interpreter/CIFactory.cpp

```diff
@@ -962,21 +1060,65 @@ static void stringifyPreprocSetting(PreprocessorOptions& PPOpts,
     // Set up the ASTContext
     CI->createASTContext();
 
-    if (OnlyLex) {
-      assert(customConsumer == nullptr && "Can't specify a custom consumer"
-                                          " when in OnlyLex mode");
-      class IgnoreConsumer: public clang::ASTConsumer {};
-      CI->setASTConsumer(
-          std::unique_ptr<clang::ASTConsumer>(new IgnoreConsumer()));
-    } else {
+    std::vector<std::unique_ptr<ASTConsumer>> Consumers;
+
+    if (!OnlyLex) {
       assert(customConsumer != nullptr && "Need to specify a custom consumer"
                                           " when not in OnlyLex mode");
-      CI->setASTConsumer(std::unique_ptr<clang::ASTConsumer>(customConsumer));
+      std::unique_ptr<clang::ASTConsumer> UniqueConsumer(customConsumer);
+      Consumers.push_back(std::move(UniqueConsumer));
     }
 
+    // With C++ modules, we now attach the consumers that will handle the
+    // generation of the PCM file itself.
+    if (COpts.CxxModules) {
+      // Code below from the (private) code in the GenerateModuleAction class.
```

Comment:
Would the use of `WriteAST()` simplify things?

Reply:
I think you were on vacation when the discussion about this happened, but to sum it up: Modules can't just be written via WriteAST(). They need a listener that we attach here that keeps track on things like what other module we load and in what order. This code wraps the WriteAST, the creation/attaching of the listener, etc. in the way with the least amount of copy pasted code from clang.

In the future we can either drop this code once I get rootcling into a form where we can reuse the whole GenerateModuleAction from clang or we make it public that we can easily reuse it. But first I want to see how far the rootcling refactor idea allows us to go :)